File: control/backend/wiim_control/playback/queue.py
"""Server-side queue engine with shuffle/repeat logic."""

import random

from wiim_control.playback.models import QueueTrack, RepeatMode, ShuffleMode


class PlayQueue:
    """Manages a play queue for a single device/group target."""

    def __init__(self):
        self.tracks: list[QueueTrack] = []
        self.position: int = 0
        self.shuffle_mode: ShuffleMode = ShuffleMode.OFF
        self.repeat_mode: RepeatMode = RepeatMode.OFF
        self._shuffle_order: list[int] = []
        self._shuffle_pos: int = 0
        # Category context: tracks belonging to current album/artist
        self._category_start: int = 0
        self._category_end: int = 0

    def set_tracks(self, tracks: list[QueueTrack], start_index: int = 0):
        """Replace the queue with new tracks."""
        self.tracks = tracks
        self.position = start_index
        self._rebuild_shuffle()

    def add_tracks(self, tracks: list[QueueTrack], next: bool = False):
        """Add tracks to the queue."""
        if next:
            insert_at = self.position + 1
            for i, track in enumerate(tracks):
                self.tracks.insert(insert_at + i, track)
        else:
            self.tracks.extend(tracks)
        self._rebuild_shuffle()

    def remove_track(self, index: int):
        if 0 <= index < len(self.tracks):
            self.tracks.pop(index)
            if self.position >= len(self.tracks):
                self.position = max(0, len(self.tracks) - 1)
            self._rebuild_shuffle()

    def reorder(self, from_index: int, to_index: int):
        if 0 <= from_index < len(self.tracks) and 0 <= to_index < len(self.tracks):
            track = self.tracks.pop(from_index)
            self.tracks.insert(to_index, track)

    def current(self) -> QueueTrack | None:
        if not self.tracks:
            return None
        idx = self._effective_index()
        if 0 <= idx < len(self.tracks):
            return self.tracks[idx]
        return None

    def advance(self) -> QueueTrack | None:
        """Move to the next track based on shuffle/repeat mode."""
        if not self.tracks:
            return None

        if self.repeat_mode == RepeatMode.TRACK:
            return self.current()

        if self.shuffle_mode != ShuffleMode.OFF:
            return self._advance_shuffled()

        # Sequential advance
        self.position += 1
        if self.position >= len(self.tracks):
            if self.repeat_mode in (RepeatMode.ALL, RepeatMode.CATEGORY):
                self.position = 0
            else:
                self.position = len(self.tracks) - 1
                return None  # end of queue

        return self.current()

    def go_back(self) -> QueueTrack | None:
        """Move to the previous track."""
        if not self.tracks:
            return None

        if self.shuffle_mode != ShuffleMode.OFF and self._shuffle_order:
            self._shuffle_pos = max(0, self._shuffle_pos - 1)
            self.position = self._shuffle_order[self._shuffle_pos]
            return self.current()

        self.position = max(0, self.position - 1)
        return self.current()

    def clear(self):
        self.tracks.clear()
        self.position = 0
        self._shuffle_order.clear()
        self._shuffle_pos = 0

    def _effective_index(self) -> int:
        if self.shuffle_mode != ShuffleMode.OFF and self._shuffle_order:
            return self._shuffle_order[self._shuffle_pos]
        return self.position

    def _advance_shuffled(self) -> QueueTrack | None:
        self._shuffle_pos += 1
        if self._shuffle_pos >= len(self._shuffle_order):
            if self.repeat_mode in (RepeatMode.ALL, RepeatMode.CATEGORY):
                self._rebuild_shuffle()
                self._shuffle_pos = 0
            else:
                return None

        self.position = self._shuffle_order[self._shuffle_pos]
        return self.current()

    def _rebuild_shuffle(self):
        if self.shuffle_mode == ShuffleMode.OFF or not self.tracks:
            self._shuffle_order = []
            return

        indices = list(range(len(self.tracks)))
        random.shuffle(indices)
        # Ensure current track is first if we have one
        if self.position in indices:
            indices.remove(self.position)
            indices.insert(0, self.position)
        self._shuffle_order = indices
        self._shuffle_pos = 0

File: backend/src/control/queue.rs
use dashmap::DashMap;
use parking_lot::RwLock;
use rand::seq::SliceRandom;
use std::sync::Arc;

use super::models::QueueTrackResponse;

/// Server-side queue for a single playback target.
#[derive(Debug)]
pub struct PlayQueue {
    tracks: Vec<QueueTrackResponse>,
    position: usize,
    shuffle_mode: String,
    repeat_mode: String,
    shuffle_order: Vec<usize>,
}

impl Default for PlayQueue {
    fn default() -> Self {
        Self::new()
    }
}

impl PlayQueue {
    pub fn new() -> Self {
        Self {
            tracks: Vec::new(),
            position: 0,
            shuffle_mode: "off".to_string(),
            repeat_mode: "off".to_string(),
            shuffle_order: Vec::new(),
        }
    }

    pub fn set_tracks(&mut self, tracks: Vec<QueueTrackResponse>, start_index: usize) {
        self.tracks = tracks;
        self.position = start_index;
        self.rebuild_shuffle_order();
    }

    pub fn add_tracks(&mut self, tracks: Vec<QueueTrackResponse>, position: &str) {
        match position {
            "next" => {
                let insert_at = (self.position + 1).min(self.tracks.len());
                for (i, t) in tracks.into_iter().enumerate() {
                    self.tracks.insert(insert_at + i, t);
                }
            }
            _ => {
                self.tracks.extend(tracks);
            }
        }
        self.rebuild_shuffle_order();
    }

    pub fn remove_track(&mut self, index: usize) -> bool {
        if index >= self.tracks.len() {
            return false;
        }
        self.tracks.remove(index);
        if self.position > 0 && index < self.position {
            self.position -= 1;
        }
        if self.position >= self.tracks.len() && !self.tracks.is_empty() {
            self.position = self.tracks.len() - 1;
        }
        self.rebuild_shuffle_order();
        true
    }

    pub fn current(&self) -> Option<&QueueTrackResponse> {
        self.tracks.get(self.position)
    }

    pub fn advance(&mut self) -> Option<&QueueTrackResponse> {
        if self.tracks.is_empty() {
            return None;
        }
        match self.repeat_mode.as_str() {
            "track" => return self.tracks.get(self.position),
            "all" => {
                self.position = (self.position + 1) % self.tracks.len();
            }
            _ => {
                if self.position + 1 < self.tracks.len() {
                    self.position += 1;
                } else {
                    return None;
                }
            }
        }
        self.tracks.get(self.position)
    }

    #[allow(dead_code)]
    pub fn go_back(&mut self) -> Option<&QueueTrackResponse> {
        if self.tracks.is_empty() {
            return None;
        }
        if self.position > 0 {
            self.position -= 1;
        } else if self.repeat_mode == "all" {
            self.position = self.tracks.len() - 1;
        }
        self.tracks.get(self.position)
    }

    pub fn clear(&mut self) {
        self.tracks.clear();
        self.position = 0;
        self.shuffle_order.clear();
    }

    pub fn tracks(&self) -> &[QueueTrackResponse] {
        &self.tracks
    }

    #[allow(dead_code)]
    pub fn tracks_mut(&mut self) -> &mut Vec<QueueTrackResponse> {
        &mut self.tracks
    }

    /// Move a track from one index to another, adjusting position accordingly.
    pub fn move_track(&mut self, from: usize, to: usize) -> bool {
        if from >= self.tracks.len() || to >= self.tracks.len() || from == to {
            return false;
        }
        let track = self.tracks.remove(from);
        self.tracks.insert(to, track);

        // Adjust current position to follow the playing track
        if self.position == from {
            // The playing track was moved
            self.position = to;
        } else if from < self.position && to >= self.position {
            // Moved a track from before position to after — shift left
            self.position -= 1;
        } else if from > self.position && to <= self.position {
            // Moved a track from after position to before — shift right
            self.position += 1;
        }
        self.rebuild_shuffle_order();
        true
    }

    pub fn position(&self) -> usize {
        self.position
    }

    pub fn shuffle_mode(&self) -> &str {
        &self.shuffle_mode
    }

    pub fn repeat_mode(&self) -> &str {
        &self.repeat_mode
    }

    pub fn set_shuffle_mode(&mut self, mode: String) {
        self.shuffle_mode = mode;
        self.rebuild_shuffle_order();
    }

    pub fn set_repeat_mode(&mut self, mode: String) {
        self.repeat_mode = mode;
    }

    fn rebuild_shuffle_order(&mut self) {
        if self.shuffle_mode == "off" || self.tracks.is_empty() {
            self.shuffle_order.clear();
            return;
        }
        let mut order: Vec<usize> = (0..self.tracks.len()).collect();
        let mut rng = rand::rng();
        order.shuffle(&mut rng);
        // Move current track to front
        if let Some(pos) = order.iter().position(|&i| i == self.position) {
            order.swap(0, pos);
        }
        self.shuffle_order = order;
    }
}

/// Manages queues for all playback targets.
pub struct QueueManager {
    queues: DashMap<String, Arc<RwLock<PlayQueue>>>,
}

impl Default for QueueManager {
    fn default() -> Self {
        Self::new()
    }
}

impl QueueManager {
    pub fn new() -> Self {
        Self {
            queues: DashMap::new(),
        }
    }

    pub fn get_or_create(&self, target_id: &str) -> Arc<RwLock<PlayQueue>> {
        self.queues
            .entry(target_id.to_string())
            .or_insert_with(|| Arc::new(RwLock::new(PlayQueue::new())))
            .clone()
    }
}
